docs/tracing.md

```diff
@@ -0,0 +1,246 @@
+# Tracing
+
+## Introduction
+
+Firecracker implements a framework for instrumentation based tracing.
+
+Instrumentation based tracing as defined by
+[Sheng Liang on usenix.org](https://www.usenix.org/legacy/publications/library/proceedings/coots99/full_papers/liang/liang_html/node9.html):
+
+> There are two ways to obtain profiling information: either statistical
+> sampling or code instrumentation. Statistical sampling is less
+> disruptive to program execution, but cannot provide completely
+> accurate information. Code instrumentation, on the other hand, may be
+> more disruptive, but allows the profiler to record all the events it
+> is interested in. Specifically in CPU time profiling, statistical
+> sampling may reveal, for example, the relative percentage of time
+> spent in frequently-called methods, whereas code instrumentation can
+> report the exact number of time each method is invoked.
+
+The focus on tracing in Firecracker is to improve debug-ability.
+
+## Implementation
+
+Firecracker implements instrumentation based tracing via
+[`log`](https://github.com/rust-lang/log) and
+[`log_instrument`](https://github.com/JonathanWoollett-Light/log-instrument),
+outputting a `Trace` level log when entering and exiting every function.
+
+It is disabled by default at compile-time, meaning the tracing
+functionality has no impact on the release binary.
+
+To enable tracing in Firecracker, compile with `--features tracing` e.g.
+`cargo build --features tracing`.
+
+This will result in an increase in the binary size (~100kb) and a
+significant regression in performance (>10x). To mitigate the
+performance impact you can filter the tracing output as described in the
+next section.
+
+## Filtering
+
+You can filter tracing output both at run-time and compile-time.
+This can be used to mitigate the performance impact of logging many
+traces.
+
+Run-time filtering is implemented with the `/logger` API call, this can
+significantly mitigate the impact on execution time but cannot mitigate
+the impact on memory usage.
```

Comment:
I've expanded it
```text
Run-time filtering is implemented with the `/logger` API call, this can
significantly mitigate the impact on execution time but cannot mitigate
the impact on memory usage. Execution time impact is mitigated by
avoiding constructing and writing the trace log, it still needs to check
the condition in every place it would output a log. Memory usage impact
is not mitigated as the instrumentation remains in the binary unchanged.
```